src/platform/datapath_raw.c

```diff
@@ -419,15 +419,122 @@ CxPlatSendDataIsFull(
+
+    //
+    // Find the best next hop IP address.
+    //
+    uint16_t SavedLocalPort = Route->LocalAddress.Ipv4.sin_port;
+    Status =
+        GetBestRoute2(
+            NULL, // InterfaceLuid
+            IFI_UNSPECIFIED, // InterfaceIndex
+            &Route->LocalAddress, // SourceAddress
+            &Route->RemoteAddress, // DestinationAddress
+            0, // AddressSortOptions
+            &IpforwardRow,
+            &Route->LocalAddress); // BestSourceAddress
+    Route->LocalAddress.Ipv4.sin_port = SavedLocalPort;
+    if (Status != ERROR_SUCCESS) {
+        QuicTraceEvent(
+            DatapathErrorStatus,
+            "[data][%p] ERROR, %u, %s.",
+            Socket,
+            Status,
+            "GetBestRoute2");
+        goto Done;
+    }
+
+    //
+    // Look up the source interface link-layer address.
+    //
+    MIB_IF_ROW2 IfRow = {0};
+    IfRow.InterfaceIndex = IpforwardRow.InterfaceIndex;
+    Status = GetIfEntry2(&IfRow);
+    if (Status != ERROR_SUCCESS) {
+        QuicTraceEvent(
+            DatapathErrorStatus,
+            "[data][%p] ERROR, %u, %s.",
+            Socket,
+            Status,
+            "GetIfEntry2");
+        goto Done;
+    }
+    CXPLAT_FRE_ASSERT(IfRow.PhysicalAddressLength == sizeof(Route->LocalLinkLayerAddress));
+    CxPlatCopyMemory(&Route->LocalLinkLayerAddress, IfRow.PhysicalAddress, sizeof(Route->LocalLinkLayerAddress));
+
+    //
+    // Map the next hop IP address to a link-layer address.
+    //
+    MIB_IPNET_ROW2 IpnetRow = {0};
+    IpnetRow.InterfaceIndex = IpforwardRow.InterfaceIndex;
+    if (QuicAddrIsWildCard(&IpforwardRow.NextHop)) { // On-link?
+        IpnetRow.Address = Route->RemoteAddress;
+    } else {
+        IpnetRow.Address = IpforwardRow.NextHop;
+    }
+    //
+    // First call GetIpNetEntry2 to see if there's already a cached neighbor. If there
+    // isn't one, or if the cached neighbor's state is unreachable (which, NB, can happen
+    // in the case where a route lookup resulted in a dummy neighbor entry being created
+    // in TCPIP.sys) or incomplete, then do a solicitation.
+    //
+    Status = GetIpNetEntry2(&IpnetRow);
+    if (Status != ERROR_SUCCESS || IpnetRow.State <= NlnsIncomplete) {
+        Status =
+            ResolveIpNetEntry2(
+                &IpnetRow,
+                &Route->LocalAddress);
+        if (Status != 0) {
+            QuicTraceEvent(
+                DatapathErrorStatus,
+                "[data][%p] ERROR, %u, %s.",
+                Socket,
+                Status,
+                "ResolveIpNetEntry2");
+            goto Done;
+        }
+    }
+    CXPLAT_FRE_ASSERT(IpnetRow.PhysicalAddressLength == sizeof(Route->NextHopLinkLayerAddress));
+    CxPlatCopyMemory(&Route->NextHopLinkLayerAddress, IpnetRow.PhysicalAddress, sizeof(Route->NextHopLinkLayerAddress));
+
+    Route->Resolved = TRUE;
+
+Done:
+
+    // TODO: convert NETIO_STATUS to QUIC_STATUS
```

Comment:
Any advice here?

Reply:
HRESULT_FROM_WIN32 should work. Since it looks like NETIO_STATUS is just a normal DWORD win32 status code.